tinytuya/Cloud.py

```diff
@@ -451,9 +458,55 @@ def getdevices(self, verbose=False):
                 ERR_CLOUD,
                 "Unable to get device list"
             )
-        else:
-            # Filter to only Name, ID and Key
-            return self.filter_devices( json_data['result'] )
+
+        self.getdevices_raw = json_data
+        devs = json_data['result']
+        changed_devices = []
+        unchanged_devices = []
+
+        # chect to see if anything has changed.  if so, re-download factory-infos and DP mapping
+        for dev in devs:
+            dev_id = dev['id']
+            if dev_id not in old_devices:
+                # newly-added device
+                changed_devices.append( dev )
+                continue
+            old = old_devices[dev_id]
+            if 'key' not in old or old['key'] != dev['local_key']:
+                # local key changed
+                changed_devices.append( dev )
+                continue
+            if (('icon' not in old) and ('icon' in dev)) or (include_map and ('mapping' not in old or old['mapping'] is None)):
+                # icon or mapping added
+                #changed_devices.append( dev )
+                #continue
+                pass

```

Comment:
Is this on purpose?

If you have an existing `devices.json` (most existing users), it will not download the DP mappings since it finds the device already there.  Suggestion we consider:

* If user requests to get DP mappings, and DEVICEFILE exists, prompt if they want to replace file or only add new devices
* Adjust this section of `getdevices()` to add mappings if missing (currently `pass`)?

Reply:
Oops, yeah, that was commented out for testing and I meant to enable it before pushing.  Fixed.

Reply:
Awesome!  

FYI, for my 41 devices `devices.json` was 23k before and now 91k.  

This data is amazing, @uzlonewolf.  Nice job.  Merging and testing....